src/ifcsverchok/__init__.py

```diff
@@ -84,7 +96,131 @@ def make_node_list():
+    filepath: bpy.props.StringProperty(subtype="FILE_PATH")
+    node_group: bpy.props.StringProperty(default="")
+    force_mode: bpy.props.BoolProperty(default=False)
+
+    @classmethod
+    def poll(cls, context):
+        return any("IFC" in n for n in context.space_data.edit_tree.nodes.keys())
+
+    def ensure_hirarchy(self, file):
+        elements_in_buildings = []
+        if not 0 <= 0 < len(file.by_type("IfcBuilding")):
+            my_building = ifcopenshell.api.run("root.create_entity", file, ifc_class="IfcBuilding", name="My Building")
+            elements = ifcopenshell.util.element.get_decomposition(my_building)
+        else:
+            for building in file.by_type("IfcBuilding"):
+                elements = ifcopenshell.util.element.get_decomposition(building)
+                elements_in_buildings.extend(elements)
+
+        for spatial in (file.by_type("IfcSpatialElement") or file.by_type("IfcSpatialStructureElement")):
+            if (not (spatial.is_a("IfcSite") or spatial.is_a("IfcBuilding")) and (spatial not in elements_in_buildings)):
+                elements = ifcopenshell.util.element.get_decomposition(spatial)
+                ifcopenshell.api.run("aggregate.assign_object", file, product=spatial, relating_object=file.by_type("IfcBuilding")[0])
 
+        elements_in_buildings_after = []
+        for building in file.by_type("IfcBuilding"):
+            elements = ifcopenshell.util.element.get_decomposition(building)
+            elements_in_buildings_after.extend(elements)
+
+        elements = file.by_type("IfcElement")
+        for element in elements:
+            if element not in elements_in_buildings:
+                ifcopenshell.api.run("spatial.assign_container", file, product=element, relating_structure=file.by_type("IfcBuilding")[0])
+
+
+        for building in file.by_type("IfcBuilding"):
+            elements = ifcopenshell.util.element.get_decomposition(building)
+            if not building.Decomposes:
+                if not 0 <= 0 < len(file.by_type("IfcSite")):
+                    ifcopenshell.api.run("root.create_entity", file, ifc_class="IfcSite", name="My Site")
+                ifcopenshell.api.run("aggregate.assign_object", file, product=building, relating_object=file.by_type("IfcSite")[0])
+                try:
+                    if file.by_type("IfcSite")[0].Decomposes[0].RelatingObject.is_a("IfcProject"):
+                        continue
+                except IndexError:
+                    pass
+                ifcopenshell.api.run("aggregate.assign_object", file, product=file.by_type("IfcSite")[0], relating_object=file.by_type("IfcProject")[0])
+        self.file = file
+        return
+
+    def execute(self, context):
+        self.file = SvIfcStore.file
+        if not self.file:
+            raise Exception("No IFC file in SvIfcStore.")
+        _, ext = splitext(self.filepath)
+        if not ext:
+            raise Exception("Bad path. Provide a path to a file.")
```

Comment:
Strictly speaking filenames do not need to have an extension. Did you meant checking something like `not os.path.isdir()`?

Reply:
Yea, I know files don't need extensions, but since there are no other checks of overwriting/validity, I made the assumption that if there is an extension the written path was conscious. Since os.path.isdir() of an 'invalid' path will return false. But I wrote this for the write_file node, where the user can actually give raw input paths. The file picker pop-up window in the panel, makes the whole operation much safer, so maybe it can be deleted/changed.